### ISSUE TITLE
Map subfolders as a top level namespace in external module resolution?

### ISSUE THREAD

**Issue by masaeedu** [TypeStrong/atom-typescript#472]
I'm using JSPM for package management, and I'm trying to find a way to get a good workflow for typescript files that utilize .d.ts files provided in libraries installed via JSPM. Basically what I want is for the editor to see `import { X, Y, Z } from 'somemodule/submodule'`, look at some configuration option in my `tsconfig.json`, and resolve `X`, `Y`, and `Z` to types defined in files at `path/to/tsconfig/.../somemodulefolder/submodule[.d].ts`. The `...` represents whatever folder structure your package manager of choice uses.

For clarity, here is a depth limited tree of my project folder:

    D:\path\to\my\project
    ├──app
    │  ├──css
    │  │  └──regr.css
    │  └──main.ts                <- import {X, Y, Z} from 'angular2/angular2'
    ├──jspm_packages
    │  ├──github
    │  │  ├──components
    │  │  ├──es-shims
    │  │  ├──jspm
    │  │  ├──mbostock
    │  │  ├──mhegazy
    │  │  ├──systemjs
    │  │  └──twbs
    │  ├──npm
    │  │  ├──angular2@2.0.0-alpha.31
	│  │  │  ...                 ┐
	│  │  │  ├──angular2.d.ts    ├── Contain definitions for X, Y, Z
	│  │  │  └──angular2.js      ┘ 
    │  │  ├──assert@1.3.0
    ...
    ├──config.js
    ├──index.html
    ├──package.json
    ├──tsconfig.json
    └──tsd.json

Certain folders in `jspm_packages/npm` and `jspm_packages/github` act as top level "namespaces" in jspm's module resolution. E.g. if I did `var angular2 = require('angular2/angular2');`, I would get the commonjs module at `jspm_packages/npm/angular2@2.0.0-alpha.31/angular2.js`. It would be excellent if the editor could be made aware of a directory root associated with a top level module, so that the import scenario described above would "just work".

**Comment by masaeedu** [comment]
One rather ugly workaround that I am using right now is symlinking the appropriate folder to the folder containing tsconfig.json. Symlinking the `angular2@2.0.0-alpha.31` folder to the root of my project under the name `angular2` results in the scenario above working correctly.

**Comment by basarat** [comment]
> symlinking the angular2@2.0.0-alpha.31 folder to the root of my project under the name angular2 results in the scenario above working correctly.

Alternatively add `jspm_packages/npm/angular**/**/*.d.ts` to your `filesGlob`

**Comment by masaeedu** [comment]
@basarat If I don't have the `angular2` symlink, the top level module `angular2` can't be resolved, and hence none of the imports within the library itself work. The problem isn't that I'm missing files (I have no `files` or `filesGlob`, so all files are included by default), the problem is that the module `angular2/angular2` doesn't resolve to `some/subfolder/relative/to/tsconfig/angular2[.d].ts`.

**Comment by basarat** [comment]
Ah ... doesn't `angular2.d.ts` contain something like `declare module "angular2/angular2"`?

**Comment by masaeedu** [comment]
That's the thing, it doesn't use absolute names. All the modules are defined with top level exports.

**Comment by masaeedu** [comment]
After doing a bit more research, I found https://github.com/Microsoft/TypeScript/issues/2338, which seems like it is proposing something similar for TypeScript itself. It isn't clear whether this has already been implemented and where the mappings are supposed to go, but it looks promising.

**Comment by basarat** [comment]
> It isn't clear whether this has already been implemented

Nope. I'm watching that issue too. 

Closing as it needs to be fixed in the compiler :rose:

**Comment by masaeedu** [comment]
@basarat It looks like this feature is still a ways off from being included in TypeScript. That issue has been closed with Node support, which doesn't help this scenario, and from what I've heard on Gitter they're waiting for the whatwg loader standard to solidify before implementing support for generic loader implementations. 

As an editor plugin, this project has a bit more leeway in terms of what features you can add, so maybe support for SystemJS (or customizable) module resolution would be a good idea? Admittedly there isn't overwhelming demand, but https://github.com/TypeStrong/atom-typescript/issues/329 seems to deal with the same issue. If you could point me to a centralized location where module resolution happens I could try to implement the path aliasing proposal from https://github.com/Microsoft/TypeScript/issues/5039.

**Comment by basarat** [comment]
>  I could try to implement the path aliasing proposal from Microsoft/TypeScript#5039.

That is the right place to track it now. Thanks for linking this  :rose:

> I could try to implement the path aliasing proposal

I tried to implement the `commonjs` module resolution with blatant disregard for code resolution logic while 2338 was in progress and I broke stuff badly (e.g. conflicts in files in node_modules vs. local files). A big issue is our limited test harness ... which means that its difficult to ensure that horrible things don't happen. Due to the complexity of this would rather see it implemented in the compiler. Thanks for your interest and patience :rose:

> if you could point me to a centralized location where module resolution happens

Just for your amusement, there is a function in langaugeServiceHost can provide `resolveModuleNames` https://github.com/TypeStrong/atom-typescript/commit/c735ec5bda95ea4afd3d98e9718b17c840addeb0#diff-0653af8be8c4c111a4007133a2d6ea7a but like I said its hard to maintain beyond other magic that I have become comfortable with, so don't work on this yourself at least not if you want to send it here :rose:

Note that we provide relative path completions that should get you a lot of the way anyways.
